ogr/ogrsf_frmts/miramon/mm_wrlayr.c

```diff
@@ -0,0 +1,7038 @@
+                               FILE_TYPE *pF3d, struct MM_ZSection *pZSection)
+{
+    if (!hMiraMonLayer)
+        return 1;
+
+    // Zsection
+    if (!hMiraMonLayer->TopHeader.bIs3d)
+    {
+        pZSection->pZDescription = nullptr;
+        return 0;
+    }
+
+    if (hMiraMonLayer->ReadOrWrite == MM_WRITTING_MODE)
+    {
+        pZSection->ZHeader.dfBBminz = STATISTICAL_UNDEF_VALUE;
+        pZSection->ZHeader.dfBBmaxz = -STATISTICAL_UNDEF_VALUE;
+    }
+
+    // ZH
+    pZSection->ZHeader.nMyDiskSize = 32;
+    pZSection->ZSectionOffset = 0;
+
+    // ZD
+    if (hMiraMonLayer->ReadOrWrite == MM_WRITTING_MODE)
+    {
+        pZSection->nMaxZDescription =
+            MM_FIRST_NUMBER_OF_VERTICES * sizeof(double);
+        if (MMInitZSectionDescription(pZSection))
+            return 1;
+    }
+    else
+    {
+        pZSection->nMaxZDescription =
+            hMiraMonLayer->TopHeader.nElemCount * sizeof(double);
+        if (MMInitZSectionDescription(pZSection))
+            return 1;
+    }
+
+    if (hMiraMonLayer->LayerVersion == MM_32BITS_VERSION)
+        pZSection->nZDDiskSize = MM_SIZE_OF_ZD_32_BITS;
+    else
+        pZSection->nZDDiskSize = MM_SIZE_OF_ZD_64_BITS;
+
+    pZSection->ZDOffset = 0;
+
+    // ZL
+    if (hMiraMonLayer->ReadOrWrite == MM_WRITTING_MODE)
+    {
+        if (MMInitFlush(&pZSection->FlushZL, pF3d,
+                        (hMiraMonLayer->nMemoryRatio != 1)
+                            ? (GUInt64)(hMiraMonLayer->nMemoryRatio * MM_25MB)
+                            : MM_25MB,
+                        &pZSection->pZL, 0, sizeof(double)))
+            return 1;
+    }
+    return 0;
+}
+
+/* -------------------------------------------------------------------- */
+/*      Layer Functions: Extensions                                     */
+/* -------------------------------------------------------------------- */
+
+/* Find the last ocurrence of pszFinalPart in pszName
+    and changes it by pszNewPart.
+ 
+    Examples of desired behaviour
+    AA.pnt -> AAT.rel
+    AA.nod -> N.~idx
+    AA.nod -> N.dbf
+    AA.nod -> N.rel
+*/
+
+static int MMChangeFinalPartOfTheName(char *pszName, size_t nMaxSizeOfName,
+                                      const char *pszFinalPart,
+                                      const char *pszNewPart)
+{
+    char *pAux, *pszWhereToFind, *pszLastFound = nullptr;
+    ;
+
+    if (!pszName || !pszFinalPart || !pszNewPart)
+        return 0;
+    if (MMIsEmptyString(pszName) || MMIsEmptyString(pszFinalPart) ||
+        MMIsEmptyString(pszNewPart))
+        return 0;
+
+    if (strlen(pszName) - strlen(pszFinalPart) + strlen(pszNewPart) >=
+        nMaxSizeOfName)
+        return 1;  // It's not possible to change the final part
+
+    pszWhereToFind = pszName;
+    while (nullptr != (pAux = strstr(pszWhereToFind, pszFinalPart)))
+    {
+        pszLastFound = pAux;
+        pszWhereToFind = pAux + strlen(pAux);
+    }
```

Comment:
This whole loop could be replaced by using strrstr instead (not the 'r' between the 2 'str', to indicate reverse): https://linux.die.net/man/3/strrstr

Reply:
I think it's only linux.
In windows it has to be created, so I prefer not to create that cause it's only needed here.

Reply:
Resolve de conversation if agreed

Reply:
> In windows it has to be created, so I prefer not to create that cause it's only needed here.

ah, you're right.
(I was confusing with another code base where we indeed have redefined strrstr() on Windows. Too bad all that code is C and not C++. We could have used std::string::rfind() otherwise)

Reply:
Yes, I'm programming a personal project about some chess stuff and I am only using C++. But MiraMon is 30 years old and al code was in C. And I am more fluently in C than in C++ so I programmed in C. Sorry. Next time I can try C++ :)